#4602: add unary ops into ttnn (export ttdnn ops through ttnn)

muthutt opened the pull request:
- roughly 200+ ops need to be exported from tt_dnn into ttnn; 
- added 62 unary ops into ttnn

Reference: (TT users only) https://docs.google.com/spreadsheets/d/115awGoFfDJUcMvuEFs7sq9Qogjogm4vWOxJwmwneTyA/edit?usp=sharing

Commit:
#0: add 62 unary ops into ttnn
  : add docs
  : update personal dict
  : permute already covered

umadevimcw committed:
#4607: Export squared difference to ttnn

umadevimcw committed:
#4602: Update the format of unary list for document

umadevimcw committed:
#4607: Add ttnn support for subalpha

umadevimcw committed:
#4607: Export hypot to ttnn

umadevimcw committed:
#4607: Export atan2 op to ttnn

umadevimcw committed:
#4607: Export Addalpha to ttnn

umadevimcw committed:
#4602: Export logical unary immediate ops to ttnn

umadevimcw committed:
#4602: Export logical ops to ttnn

umadevimcw committed:
#4607: Export ldexp, lodaddexp and logaddexp2 to ttnn

umadevimcw committed:
#4602: Export rdiv, rpow and rsub to ttnn

umadevimcw committed:
#4607: Export nextafter and outer op to ttnn

VirdhatchaniKN committed:
#4604: Export addcmul, addcdiv, where, mac to ttnn

bharane-ab committed:
#4605: Export relational operators to ttnn

umadevimcw committed:
#4607: Export xlogy to ttnn

Commit:
#4606: Complex op Support in TTNN
     : update complex ops export into ttnn with unittests;
       - TODO: ttnn.polar and ttnn.angle need some checks on accuracy

arakhmati reviewed (changes requested):
Can this PR please be broken down into 4 PRs?
- unary ops
- binary ops
- ternary ops
- relational ops

Commit:
#0: add norm, std, mean, var functions to unary with reduce bent to it
  : update test files to tests/ttnn/ directory
  : move test

Commit:
#0: add creation ops - ones, zeros, full and variants based off the existing tensors
    :

umadevimcw committed:
#4607: Update ttnn binary test with comp

umadevimcw committed:
#4607: Update test of binary ops

umadevimcw committed:
#4607: Export assign op to ttnn

umadevimcw committed:
#4602: Export heaviside op to ttnn

umadevimcw committed:
#4607: Export isclose op to ttnn

umadevimcw committed:
#4607: Export loss function to ttnn

bharane-ab committed:
#4605: Export relational operator to ttnn

Commit:
#0: add more passing tests to dunder relational
  : update relational tests to run them correctly for correspondence between actual and golden
  : update spellcheck dictionary

davorchap commented:
> Can this PR please be broken down into 4 PRs?
> 
> * unary ops
> * binary ops
> * ternary ops
> * relational ops

Yes. And the docs should also be structured into MM/Conv , TMs, Unary, Binary, Ternary, Relational,  TMs, Transformer

eyonland commented on tests/ttnn/unit_tests/test_ternary.py:
```diff
@@ -0,0 +1,127 @@
+        output = ttnn.where(a, b, c)
+        output = ttnn.to_layout(output, ttnn.ROW_MAJOR_LAYOUT)
+
+        print(f"shape: {output.shape}")
```
I'm not seeing asserts anywhere on these tests.   It's like the test_<category of op>.py doesn't allow much real testing on these ops.

eyonland commented on ttnn/binary.py:
```diff
@@ -0,0 +1,205 @@
+    ) -> Tensor:
+        f"""{name}(input_tensor_a: Tensor, input_tensor_b: Tensor) -> Tensor
+
+        Applies {name} to :attr:`input_tensor_a` and  :attr:`input_tensor_b` element-wise.
```
Looks like the documentation is unique per category of op.   This means that it will not be possible to easily add custom documentation that expands the definition.

umadevimcw committed:
#4607: Remove outer

umadevimcw committed:
#4607: Export min and max op

Commit:
#0: Fix polar and angle op

umadevimcw committed:
#4601: Export gelu and silu based activation function to ttnn

bharane-ab committed:
#0: Add test for relational ttnn ops

umadevimcw committed:
#4607: Include outer with the fix. Test is running without any issue

VirdhatchaniKN committed:
#0: Export tilize_with_zero_padding, clip to ttnn, Update ttnn ternary test

umadevimcw committed:
#4607: Update op name

Commit:
#0: Tensor object has bound methods from unary, binary and ternary ops for ergonomics

Commit:
#0: move to new package area

Commit:
#0: move ttnn package into the ttnn/ttnn area
  : update package location for docs
  : update PYTHONPATH for ttnn rename package
  : fix PYTHONPATH

aliuTT commented:
You recommited a few commits from main, should be gone when you rebase

Commit:
#0: common wrapper updated

arakhmati commented:
> > Can this PR please be broken down into 4 PRs?
> > 
> > * unary ops
> > * binary ops
> > * ternary ops
> > * relational ops
> 
> can I know why ? Our reason to do it this way is b/c the changes are not affecting the infra we are just exporting the tests; since all functions need to be in one place a single PR avoids conflicts in development which are handled here already pulling in all the works from multiple people

The reason is because the whole point of `ttnn` is making better APIs than `tt_lib`. And making them consistent and user-friendly. And porting everything into `ttnn` that we have in `tt_lib` at once is not the goal. We need smaller PRs so that @eyonland and I have time to take a closer look at everything that is getting modified and added

Commit:
#0: add asserts to the relational test
  : rename the test
  : add assertions
  : all new tests have assertions.

Try to add asserts for all new tests

tests/ttnn/unit_tests/test_binary.py
tests/ttnn/unit_tests/test_complex.py
tests/ttnn/unit_tests/test_mae_loss.py
tests/ttnn/unit_tests/test_mse_loss.py
tests/ttnn/unit_tests/test_outer.py
tests/ttnn/unit_tests/test_relational.py
tests/ttnn/unit_tests/test_relational_2.py
tests/ttnn/unit_tests/test_tensor.py
tests/ttnn/unit_tests/test_ternary.py
tests/ttnn/unit_tests/test_unary.py

Commit:
#0: Add the is_real and is_imag op

umadevimcw committed:
#4607: Export concat op to ttnn

muthutt committed:
#0: fix pythonpath cd - bad idea

muthutt committed:
#0: add ergonomic fluent interface for Tensor with tests
  : undo some bits

muthutt committed:
#0: undo ttnn/ttnn b/c this is going to take much longer

muthutt committed:
Revert "#0: fix pythonpath cd - bad idea"

This reverts commit 0491dd358c57a47a655928f8ff170ee807ec85fe.

muthutt committed:
#0: undo ttnn/ttnn package directory change

muthutt committed:
#0: update docs and make ComplexTensor more similar to Tensor

muthutt committed:
#0: fix

muthutt committed:
#0: undo change to setup.py
